dozer-types/src/models/connection.rs

```diff
@@ -16,30 +20,175 @@ pub struct Connection {
+    pub connection_url: Option<String>,
+}
+
+#[derive(Eq, PartialEq, Clone, Debug, Hash)]
+pub struct PostgresConfigReplenished {
     pub user: String,
-    #[prost(string, tag = "2")]
     pub password: String,
-    #[prost(string, tag = "3")]
     pub host: String,
-    #[prost(uint32, tag = "4")]
     pub port: u32,
-    #[prost(string, tag = "5")]
     pub database: String,
+    pub ssl_mode: SslMode,
 }
 
-impl PostgresConfig {
+impl PostgresConfigReplenished {
     pub fn convert_to_table(&self) -> Table {
         table!(
-            ["user", self.user.as_str()],
+            ["user", self.user],
             ["password", "*************"],
             ["host", self.host],
             ["port", self.port],
-            ["database", self.database]
+            ["database", self.database],
+            ["sslmode", format!("{:?}", self.ssl_mode)]
         )
     }
 }
 
+impl PostgresConfig {
+    pub fn replenish(&self) -> PostgresConfigReplenished {
+        if self.connection_url.is_none() {
+            PostgresConfigReplenished {
+                user: self.user.clone().unwrap(),
+                password: self.password.clone().unwrap(),
+                host: self.host.clone().unwrap(),
+                port: self.port.unwrap(),
+                database: self.database.clone().unwrap(),
+                ssl_mode: get_ssl_mode(self.sslmode.clone()),
+            }
+        } else {
+            let map = connection_url_map(self.connection_url.as_ref().unwrap(), self);
+            PostgresConfigReplenished {
+                user: map
+                    .get("user")
+                    .expect("user is missing from connection url")
+                    .to_string(),
+                password: map
+                    .get("password")
+                    .expect("password is missing from connection url")
+                    .to_string(),
+                host: map
+                    .get("host")
+                    .expect("host is missing from connection url")
+                    .to_string(),
+                port: u32::from_str(
+                    map.get("port")
+                        .expect("port is missing from connection url"),
+                )
+                .unwrap(),
+                database: map
+                    .get("database")
+                    .expect("database is missing from connection url")
+                    .to_string(),
+                ssl_mode: get_ssl_mode(map.get("sslmode").cloned()),
+            }
+        }
+    }
+}
+
+fn get_ssl_mode(mode: Option<String>) -> SslMode {
+    match mode {
+        Some(m) => match m.as_str() {
+            "disable" | "Disable" | "" => SslMode::Disable,
+            "prefer" | "Prefer" => SslMode::Prefer,
+            "require" | "Require" => SslMode::Require,
+            &_ => SslMode::Disable,
+        },
+        None => SslMode::Disable,
+    }
+}
+
+pub fn connection_url_map(url: &str, config: &PostgresConfig) -> HashMap<String, String> {
```

Comment:
I'm not able to read this regex. Is there a library for parsing the connection string?

Reply:
> Configuration can be parsed from libpq-style connection strings

https://docs.rs/tokio-postgres/latest/tokio_postgres/config/struct.Config.html#impl-FromStr-for-Config